plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -287,14 +300,45 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
         variableAssignment: FirVariableAssignment,
         data: StmtConversionContext<ResultTrackingContext>,
     ): Exp {
+
         // It is not entirely clear whether we can get away with ignoring the distinction between
         // lvalues and rvalues, but let's try to at first, and we'll fix it later if it turns out
         // not to work.
-        val convertedLValue = data.convert(variableAssignment.lValue)
-        val lValueType = data.embedType(variableAssignment.lValue)
-        val convertedRValue = data.convert(variableAssignment.rValue)
-        val rValueType = data.embedType(variableAssignment.rValue)
-        data.addStatement(Stmt.assign(convertedLValue, convertedRValue.convertType(rValueType, lValueType)))
+
+        if (variableAssignment.isAssigningToClassField) {
+
+            val lValue = variableAssignment.lValue as FirPropertyAccessExpression
+            val lValueSymbol = lValue.calleeSymbol as FirPropertySymbol
+            val lValueType = data.embedType(variableAssignment.lValue)
+            val receiver = data.convert(lValue.dispatchReceiver)
+
+            val convertedRValue = data.convert(variableAssignment.rValue)
+            val rValueType = data.embedType(variableAssignment.rValue)
+
+            // According to the setter symbol, we should invoke a method, or assigning a value
+            val setterSymbol = lValueSymbol.setterSymbol!!
+            if (setterSymbol.hasBody) {
+                val method = data.embedFunction(setterSymbol)
+                data.withResult(UnitTypeEmbedding) {
+                    data.addStatement(method.toMethodCall(listOf(receiver, convertedRValue), this.resultCtx.resultVar))
+                }
+            } else {
+                val varEmbedding = VariableEmbedding(lValueSymbol.callableId.embedName(), lValueType)
+                val fieldAccess = Exp.FieldAccess(receiver, varEmbedding.toField())
+                val accPred = AccessPredicate.FieldAccessPredicate(fieldAccess, PermExp.FullPerm())
+                data.addStatement(Stmt.Inhale(accPred))
+                data.addStatement(Stmt.assign(fieldAccess, convertedRValue.convertType(rValueType, lValueType)))
+                data.addStatement(Stmt.Exhale(accPred))
+            }
+
+        } else {
+            val convertedLValue = data.convert(variableAssignment.lValue)
+            val lValueType = data.embedType(variableAssignment.lValue)
+            val convertedRValue = data.convert(variableAssignment.rValue)
+            val rValueType = data.embedType(variableAssignment.rValue)
```

Comment:
You are computing those values in both branches so they could be moved in front of the if.